ENYO-493: Duration Picker(hour, minute, second)

anishlg opened the pull request:
updated the old duration picker
https://github.com/enyojs/moonstone/pull/1553

review comments updated
1. Used single value property than using hour, minute and seconds values separately.
2. Min/max values are removed as it's always constant.
3. Combined DurationPickerBase and DurationPcker into a single file as there won't be any future sub kinds for DurationPcker 
4.Made the template property published.
5. Removed DurationPickerBase.openChanged method as it doesn't do anything
6. Updated the sample.

otisnate commented on samples/DurationPickerSample.html:
```diff
@@ -0,0 +1,27 @@
+	<!-- -->
+	<meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">
+	<!-- Less.js (uncomment for client-side rendering of less stylesheets; leave commented to use only CSS)
+	<script src="../../../enyo/tools/minifier/node_modules/less/dist/less-1.7.0.min.js"></script> -->
```
Please update less related routine. You need to refer https://github.com/enyojs/moonstone/blob/master/samples/ButtonSample.html

otisnate commented on samples/DurationPickerSample.html:
```diff
@@ -0,0 +1,27 @@
+	<script src="../../../enyo/tools/minifier/node_modules/less/dist/less-1.7.0.min.js"></script> -->
+	<script src="../../../enyo/enyo.js" type="text/javascript"></script>
+	<script src="../../enyo-ilib/package.js" type="text/javascript"></script>
+	<script src="../../enyo-cordova/package.js" type="text/javascript"></script>
```
Do you need cordova APIs? If not, you can remove this line

otisnate commented:
Please update manifest.json and manifest-light.json file. Then you need to verify if Sampler shows Duration Picker sample

Djspaceg commented on samples/DurationPickerSample.js:
```diff
@@ -0,0 +1,51 @@
+		{kind: 'moon.InputDecorator', components: [
+			{kind: 'moon.Input', placeholder: 'h:m:s', name: 'newTemplate'}			
+		]},
+		{kind: 'moon.Button', name: 'setTemplate', content: 'Set Template', ontap: 'setNewTemplate'},
```
Lets set these buttons to `small: true` so their height matches the inputs.

Djspaceg commented on source/DurationPicker.js:
```diff
@@ -0,0 +1,565 @@
+	*/
+
+	enyo.kind(
+		/** @lends moon.DurationPickerBase.prototype */ {
```
Docs still mention DurationPickerBase.

anishlg commented:
Duration picker is updated
1. removed the duration picker base component.
2. updated with review comments. 
3. added 'disablepicker' option to duration picker.
4. updated the duration picker sample with full, disabled and non-editable duration pickers

otisnate commented on source/DurationPicker.js:
```diff
@@ -0,0 +1,516 @@
+		*/
+		bindings: [
+			{from: '.disabled', to: '.$.headerWrapper.disabled'},
+			{from: '.disabledPicker', to: '.$.hour.disabled'},
```
Is there requirement for ```disabledPicker``` option?

Djspaceg replied:
Yes, most controls should have a disabled state.

otisnate replied:
Yes, you are right. But ```disabledPicker``` is new flag and is not same with ```disabled```

otisnate commented on source/DurationPicker.js:
```diff
@@ -0,0 +1,516 @@
+		* @private
+		*/
+		constructor: function() {
+			this.inherited(arguments);
```
Please use ```enyo.inherit``` instead of ```this.inherited``` for every functon

Djspaceg replied:
While `enyo.inherit` is technically faster, `this.inherited` makes the code easier to read. We could, during minification for example, automatically upgrade each `this.inherited` to a faster `enyo.inherit` method, because it's so consistent and easily readable. There won't be hundreds of these stamped out at a time on a single view, so the speed benefit is minimal. Either of these options is technically fine though.

otisnate commented on source/DurationPicker.js:
```diff
@@ -0,0 +1,516 @@
+		* @fires moon.DurationePicker#onChange
+		* @private
+		*/
+		handleChangeEvent: function (sender, ev) {
```
Please change ```(inSender, inEvent)``` instead of ```(sender, ev)```

Djspaceg replied:
`(inSender, inEvent)` is the old naming convention we no longer use. `(sender, ev)` is the current, shorter names.

aarontam replied:
What @Djspaceg said is correct, and just to add on to that, as a general change we have been removing the `in` prefix from all of our parameter names in favor of shorter names.

otisnate replied:
When can we check current shorter name's rule?

Djspaceg replied:
This change of standards was made in July of 2014, during the docs migration and code cleanup.

otisnate commented on samples/DurationPickerSample.js:
```diff
@@ -0,0 +1,73 @@
+			this.$.result.set('content', ev.name + ' expired');
+		}
+	},
+	resetTapped: function(sender, ev) {
```
Please add ```this.$.pickerDuration.set("open", false);```

Djspaceg replied:
Why? Shouldn't you be able to reset the picker with it still open? This is a sample.

aarontam replied:
I *think* there is a (known?) issue where resetting a picker while it is still open results in the child picker values not being updated until the picker is closed and opened again. You can see this in the `DatePickerSample` as well, which has the same behavior and does not contain an explicit call to close the picker. It's possible that the explicit call to close the picker was an app workaround for this, but I'm unclear on how frequent the case is to reset the picker while it is open. In any case, this might be something to dig into to fix, as a separate issue, and we probably wouldn't want the explicit call to close the picker since it would just be a workaround and not ideal.

Djspaceg replied:
This behavior doesn't seem to be the case with this picker in my local tests. Are you seeing otherwise?

aarontam replied:
Sorry just saw this (didn't get a notification for some reason)! I hadn't had a chance to actually try out the DurationPicker at the time - it appears to be working correctly for the case of resetting, so it was a workaround that was specific to `moon.DateTimePickerBase`, which is probably context for @otisnate's request/comment. In any case, it doesn't look like we need to add the call here.

otisnate replied:
I just want to sync up with TimePickerSample - https://github.com/enyojs/moonstone/blob/master/samples/TimePickerSample.js#L77

anishlg commented:
The duration picker has updated
1. add start/cancel and pause/resume buttons.
2. hide/show button option for application developer.
3. the application developer can set any order for hour, minute and second
4. the picker values are removed 0 from single digits

aarontam commented on source/DurationPicker.js:
```diff
@@ -0,0 +1,733 @@
+		* @private
+		*/
+		bindings: [
+			{from: '.disabled', to: '.$.headerWrapper.disabled'},
```
Hi @anishlg, per @Djspaceg's previous comment, we can and should remove the leading "." before each of these binding paths as they are no longer necessary, and the removal of the "." prefix will speed things up.

aarontam commented on source/DurationPicker.js:
```diff
@@ -0,0 +1,733 @@
+		*/
+		closePicker: function (sender, ev) {
+			/**
+			* If select/enter is pressed on any date picker item or the left key is pressed on the
```
Pressing 5-way left on the first item does not seem to close the drawer.

aarontam commented on source/DurationPicker.js:
```diff
@@ -0,0 +1,733 @@
+		/**
+		* @private
+		*/
+		hourTextChanged: function (inOldvalue, inNewValue) {
```
Can we update the parameter names so that they do not use the "in" prefix here? Similarly for the other two change handlers that follow.

aarontam commented on source/DurationPicker.js:
```diff
@@ -0,0 +1,733 @@
+		* @private
+		*/
+		changeCountdown: function () {
+			if (this.countdown) {
```
Can we simplify this entire method to:

```javascript
if (this.$.hour)
	this.$.hour.set('disabled', this.countdown);
if (this.$.minute)
	this.$.minute.set('disabled', this.countdown);
if (this.$.second)
	this.$.second.set('disabled' , this.countdown);
```

aarontam commented on source/DurationPicker.js:
```diff
@@ -0,0 +1,733 @@
+		* @default moon.$L('Start')
+		* @private
+		*/
+		startConent: moon.$L('Start'),
```
This should probably be `startContent`.

aarontam commented on source/DurationPicker.js:
```diff
@@ -0,0 +1,733 @@
+		* @default moon.$L('Pause')
+		* @private
+		*/
+		pauseConent: moon.$L('Pause'),
```
This should probably be `pauseContent`.

aarontam commented on source/DurationPicker.js:
```diff
@@ -0,0 +1,733 @@
+		* @private
+		*/
+		observers: {
+			'changeCountdown': ['countdown']
```
Do we need to explicitly specify this observer, or could we instead just implement a `countdownChanged` method?

aarontam commented on source/DurationPicker.js:
```diff
@@ -0,0 +1,733 @@
+		* @default {}
+		* @private
+		*/
+		valueArray: {},
```
This is on the nitpicky side, but this is an object and not an array. :)

aarontam commented on source/DurationPicker.js:
```diff
@@ -0,0 +1,733 @@
+		*/
+		constructor: function () {
+			this.inherited(arguments);
+			this.valueArray = {
```
Is there a specific reason we need to initialize these here, as opposed to doing so in `create`?

aarontam commented on source/DurationPicker.js:
```diff
@@ -0,0 +1,733 @@
+		/**
+		* @private
+		*/
+		currentTemplate: function () {
```
On the nitpicky side again, but this sounds more like a property name than a method - how about something like 'initTemplate` or `parseTemplate`?

aarontam commented on source/DurationPicker.js:
```diff
@@ -0,0 +1,733 @@
+		* @private
+		*/
+		setupArray: function () {
+			if (this.template.search('h') < 0)
```
I think using `indexOf` would be more performant here: http://jsperf.com/exec-vs-match-vs-test-vs-search/11.

aarontam commented on source/DurationPicker.js:
```diff
@@ -0,0 +1,733 @@
+					item = this.templateArray[idx];
+					switch (item) {
+						case 'h':
+							text = text + ' ' + this.valueArray['hour'] + ' ' + moon.$L('Hours');
```
We need to take into account number agreement between the value and the units (i.e. 1 Hour).

aarontam commented on source/DurationPicker.js:
```diff
@@ -0,0 +1,733 @@
+		* @private
+		*/
+		startTapped: function () {
+			if (this.startConent == moon.$L('Start')) {
```
It seems like a bad idea to have the string value for `startContent` effectively defined in two locations as we're using it in the conditional here. Can we instead check to see if we are currently counting down, to determine which path to take?

aarontam commented on source/DurationPicker.js:
```diff
@@ -0,0 +1,733 @@
+		* @private
+		*/
+		pauseTapped: function () {
+			if (this.pauseConent == moon.$L('Pause')) {
```
Same as above - makes me uncomfortable to be comparing against an effectively static string as a way of knowing our current state.

aarontam commented on source/DurationPicker.js:
```diff
@@ -0,0 +1,733 @@
+		* @private
+		*/
+		cancelPicker: function () {
+			this.set('startConent', moon.$L('Start'));
```
In fact, we might want to consider storing these strings in a single location, possibly as private properties of this control, so we can update them all in one place. As it stands now, we would have to remember to update the text content for "Pause" in at least 4 locations, if we ever wanted to change it.

Djspaceg replied:
If you add these as locally scoped private variables, on around line#2, we'll get all of the benefits of the strings stored in one place, but none of the memory or loading overhead of a published property. Ex:
```
var pauseStr = $L('Pause),
    startStr = $L('Start');
```

aarontam replied:
:+1:

anishlg commented:
The duration picker updated with review comments.
1. removed the '.' prefix from binding
2. removed 'in' prefix from 'valueChanged()' functions
3. updated the 'countdownChanged()' fucntion
4. 'startContent' and 'pauseContent' text updated
5. removed the observer
6. replaced 'valueArray' to 'values'
7. replaced 'currentTemplate()' to 'initTemplate()'. 
8. moved the 'constructor()' code to 'initDefault()' 
9. replaced 'search()' array with 'indexOf()'
10.fixed the number agreement between the value and the units
11. created private variables for button text with ilib support and removed the direct string checking used in condition fucntions
12. resolved 1 second delay issue.

aarontam commented on source/DurationPicker.js:
```diff
@@ -0,0 +1,781 @@
+					item = this.templateArray[idx];
+					switch (item) {
+						case 'h':
+							text = text + ' ' + this.values['hour'] + ' ' + (this.values['hour'] == 1 ? moon.$L('Hour') : moon.$L('Hours'));
```
Hi @anishlg, instead of using `moon.$L('Hour')` and `moon.$L('Hours')` here, can we instead utilize our `hourText` property? Similarly, we would want to do the same thing for minutes and seconds.

aarontam commented on source/DurationPicker.js:
```diff
@@ -0,0 +1,781 @@
+		* @default moon.$L('Start')
+		* @private
+		*/
+		startText: moon.$L('Start'),			// i18n 'Start' label in moon.DurationPicker buttons
```
Having these as properties on the kind definitely works - I was curious if you had tried @Djspaceg's approach as a more efficient alternative: https://github.com/enyojs/moonstone/pull/1978#discussion_r25743288.

anishlg commented:
updated the duration picker to reuse the text property

anishlg commented:
Duration picker updated
1. the optimization suggested is implemented. all the texts are made to local private variables.
2. implemented an interval feature than using window.setInterval() as this causes time variation in javascript.

anishlg commented:
the tickInterval() renamed as moon.tickInterval() to solve the travis error

anishlg commented:
updated the duration picker
1. used fixed LTR.
2. used ilib lib for capitalized and the non-capitalized strings
3. used the ilib.DurFmt to get the formatted string

aarontam commented:
@anishlg when you next make an update, can you also fix the merge conflict? Thanks!

anishlg commented:
created new PR request 
https://github.com/enyojs/moonstone/pull/2031

hence closing this request
